README
mafw-lastfm
-----------

mafw-lastfm is a small daemon that scrobbles your listening habits to last.fm.

setting up
----------

As of mafw-lastfm 0.0.2 a control panel applet for Maemo is available. In order
to set up mafw-lastfm, simply access Settings -> Last.fm settings.

Alternatively, you can create a file $HOME/.osso/mafw-lastfm with the
following contents:

       [Credentials]
       username=[your lastfm username]
       password=[a md5sum of your lastfm password]

For instance, for a user named 'jamesthehacker' with password 'james':

	[Credentials]
	username=jamesthehacker
	password=b4cc344d25a2efe540adbf2678e2304c

To generate the md5sum of your password, you can simply run

   $ echo -n password | md5sum


project page and source packages
--------------------------------

mafw-lastfm has a garage.maemo.org page, where you can find the latest releases:

  https://garage.maemo.org/projects/mafw-lastfm/

mailing list
------------

mfaw-lastfm development (and users, why not) mailing list is <mafw-lastfm-devel@garage.maemo.org>.
You can suscribe in

  https://garage.maemo.org/mailman/listinfo/mafw-lastfm-devel

bug reports
-----------

You can report bugs in

  https://bugs.maemo.org/enter_bug.cgi?product=mafw-lastfm

source code
-----------

The source code is kept in a gitorious project. The web interface is

    http://gitorious.org/mafw-lastfm

You can clone it from

    git clone git://gitorious.org/mafw-lastfm/mafw-lastfm.git


limitations
-----------

mafw-lastfm is in an early stage and has plenty of limitations:

- The daemon won't work properly behind a proxy.
- Libre.fm is not yet supported.

These features will come with time.

audioscrobbler
--------------

The audioscrobbler icon is part of the sources of the Last.fm player.


NEWS
[0.0.8 Release]

- Rewrite of the core scrobbler code, making it more robust and
  lightweight. As a result, also:
- Scrobble tracks played long enough even if the scrobbler is killed
  before the track finishes.
- Do not scrobble a track that's paused before played long enough and
  never resumed.
- --enable-debug configure feature to enable console logging (now
    disabled by default).
- New and updated translations:
  [fi] Oskari Timperi

[0.0.7 Release]

- Store tracks on-disk. Now scrobbling survives reboots.
- Localization support.
- Plug another leak.
- New and updated translations:
  [de] Andre Klapper
  [es] Claudio Saavedra
  [en_GB] David King
  [ru] Ekaterina Gerasimova

[0.0.6 Release]

- Plug a very nasty leak.

[0.0.5 Release]

- Do not lose cached tracks after returning from offline.

[0.0.4 Release]

- Monitor credentials file, no need for reboot!
- Several fixes to handshaking after offline periods.
- Less polling during now-playing submissions.
- Code fixes and cleanups.

[0.0.3 Release]

- Credentials file should be stored now in $HOME/.osso/mafw-lastfm instead.
- Improved network error recovery.
- Use audioscrobbler icon in control panel applet.
- Work properly with broken mafw versions (track length issue).
- Less wake-ups, several code fixes.

[0.0.2 Release]

- New control panel applet for the configuration of the last.fm account.
- Handhsake after BADSESSION responses, allowing the daemon to recover
  after another client takes the session.
- Don't scrobble a paused track.
- Code fixes, leaks plugged, less wake-ups.

[0.0.1 Release]

- Initial release
- Set now-playing, and scrobble.
- Credentials stored in $HOME/.mfaw-lastfm


mafw-lastfm/mafw-lastfm.c
/**
 * mafw-lastfm: a last.fm scrobbler for mafw
 *
 * Copyright (C) 2009-2010  Claudio Saavedra <csaavedra@igalia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <glib.h>
#include <libmafw/mafw.h>
#include <libmafw-shared/mafw-shared.h>
#include <gio/gio.h>
#include <string.h>

#include "mafw-lastfm-scrobbler.h"

#define WANTED_RENDERER "Mafw-Gst-Renderer"
#define MAFW_LASTFM_CREDENTIALS_FILE ".osso/mafw-lastfm"

gint64 length;
glong current_time;
gint position;

static gchar *
mafw_metadata_lookup_string (GHashTable *table,
                             const gchar *key)
{
  GValue *value;
  value = mafw_metadata_first (table, key);
  return value ?  g_value_dup_string (value) : NULL;
}

static int
mafw_metadata_lookup_int (GHashTable *table,
                          const gchar *key)
{
  GValue *value;
  value = mafw_metadata_first (table, key);
  return value ? g_value_get_int (value) : 0;
}

static void
metadata_callback (MafwRenderer *self,
                   const gchar *object_id,
                   GHashTable *metadata,
                   gpointer user_data,
                   const GError *error)
{
  MafwLastfmScrobbler *scrobbler;
  MafwLastfmTrack *track;

  scrobbler = MAFW_LASTFM_SCROBBLER (user_data);

  track = mafw_lastfm_track_new ();

  track->artist = mafw_metadata_lookup_string (metadata, MAFW_METADATA_KEY_ARTIST);
  track->title = mafw_metadata_lookup_string (metadata, MAFW_METADATA_KEY_TITLE);

  if (!track->artist || !track->title)
    return;

  track->timestamp = current_time;
  track->source = 'P';
  track->album = mafw_metadata_lookup_string (metadata, MAFW_METADATA_KEY_ALBUM);
  track->number = mafw_metadata_lookup_int (metadata, MAFW_METADATA_KEY_TRACK);
  track->length = length;

  mafw_lastfm_scrobbler_enqueue_scrobble (scrobbler, track, position);

  mafw_lastfm_track_free (track);
}

static void
position_callback (MafwRenderer *renderer,
                   gint current_position,
                   gpointer user_data,
                   const GError *error)
{
  position = current_position;
  mafw_renderer_get_current_metadata (renderer,
                                      metadata_callback,
                                      user_data);
}

static void
state_changed_cb (MafwRenderer *renderer,
                  MafwPlayState state,
                  gpointer user_data)
{
  GTimeVal time_val;
  switch (state) {
  case Playing:
    g_get_current_time (&time_val);
    current_time = time_val.tv_sec;
    mafw_renderer_get_position (renderer, position_callback,
                                user_data);
    break;
  case Paused:
    mafw_lastfm_scrobbler_suspend (MAFW_LASTFM_SCROBBLER (user_data));
    break;
  case Stopped:
    mafw_lastfm_scrobbler_flush_queue (MAFW_LASTFM_SCROBBLER (user_data));
    break;
  default:
    break;
  }
}

static void
metadata_changed_cb (MafwRenderer *renderer,
                     gchar *name,
                     GValueArray *varray,
                     gpointer user_data)
{
  if (strcmp (name, "duration") == 0)
    length = g_value_get_int64 (g_value_array_get_nth (varray, 0));
}

static void
renderer_added_cb (MafwRegistry *registry,
                   GObject *renderer,
                   gpointer user_data)
{
  const gchar *name;

  if (!MAFW_IS_RENDERER (renderer))
    return;

  name = mafw_extension_get_name (MAFW_EXTENSION (renderer));

  if (strcmp (name, WANTED_RENDERER) != 0)
    return;

  g_signal_connect (renderer,
                    "state-changed",
                    G_CALLBACK (state_changed_cb),
                    user_data);
  g_signal_connect (renderer,
                    "metadata-changed",
                    G_CALLBACK (metadata_changed_cb),
                    user_data);
}

static gboolean
get_credentials (gchar *file,
                 gchar **username,
                 gchar **pw_md5)
{
  GKeyFile *keyfile;
  GError *error = NULL;

  keyfile = g_key_file_new ();

  if (!g_key_file_load_from_file (keyfile, file, G_KEY_FILE_NONE, &error)) {
    if (error) {
      g_warning ("Error loading credentials file: %s",
                 error->message);
      g_error_free (error);
    }

    g_key_file_free (keyfile);

    return FALSE;
  }

  *username = g_key_file_get_string (keyfile,
                                     "Credentials", "username", NULL);
  *pw_md5 = g_key_file_get_string (keyfile,
                                   "Credentials", "password", NULL);

  if (!*username || !*pw_md5) {
    g_warning ("Error loading username or password md5");

    g_free (*username);
    g_free (*pw_md5);
    g_key_file_free (keyfile);

    return FALSE;
  }

  g_key_file_free (keyfile);

  return TRUE;
}

static void
authenticate_from_file (MafwLastfmScrobbler *scrobbler,
                        gchar *path)
{
  gchar *username, *md5passwd;

  if (!get_credentials (path, &username, &md5passwd))
    return;

  mafw_lastfm_scrobbler_set_credentials (scrobbler, username, md5passwd);
  mafw_lastfm_scrobbler_handshake (scrobbler);
  g_free (username);
  g_free (md5passwd);
}

static void
on_credentials_file_changed (GFileMonitor *monitor,
                             GFile *file,
                             GFile *other_file,
                             GFileMonitorEvent event_type,
                             MafwLastfmScrobbler *scrobbler)
{
  gchar *path;

  path = g_file_get_path (file);
  authenticate_from_file (scrobbler, path);
  g_free (path);
}

static void
monitor_credentials_file (const gchar *path,
                          MafwLastfmScrobbler *scrobbler)
{
  GFile * file;
  GFileMonitor *monitor;

  file = g_file_new_for_path (path);
  monitor = g_file_monitor_file (file, G_FILE_MONITOR_NONE,
                                 NULL, NULL);
  g_signal_connect (monitor, "changed",
                    G_CALLBACK (on_credentials_file_changed),
                    scrobbler);
  g_object_unref (file);
}

int main (void)
{
  GError *error = NULL;
  MafwRegistry *registry;
  GMainLoop *main_loop;
  MafwLastfmScrobbler *scrobbler;
  gchar *file;

  g_type_init ();
  if (!g_thread_supported ())
    g_thread_init (NULL);

  scrobbler = mafw_lastfm_scrobbler_new ();

  registry = MAFW_REGISTRY (mafw_registry_get_instance ());
  if (!registry) {
    g_warning ("Failed to get register.\n");
    return 1;
  }

  mafw_shared_init (registry, &error);
  if (error) {
    g_warning ("Failed to initialize the shared library.\n");
    return 1;
  }

  g_signal_connect (registry,
                    "renderer-added",
                    G_CALLBACK (renderer_added_cb), scrobbler);

  file = g_build_filename (g_get_home_dir (),
                           MAFW_LASTFM_CREDENTIALS_FILE, NULL);
  monitor_credentials_file (file, scrobbler);
  authenticate_from_file (scrobbler, file);
  g_free (file);

  main_loop = g_main_loop_new (NULL, FALSE);
  g_main_loop_run (main_loop);

  return 0;
}


mafw-lastfm/mafw-lastfm-scrobbler.h
/**
 * mafw-lastfm: a last.fm scrobbler for mafw
 *
 * Copyright (C) 2009  Claudio Saavedra <csaavedra@igalia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef MAFW_LASTFM_SCROBBLER_H
#define MAFW_LASTFM_SCROBBLER_H

#include <glib-object.h>

G_BEGIN_DECLS

#define MAFW_LASTFM_TYPE_SCROBBLER mafw_lastfm_scrobbler_get_type ()

#define MAFW_LASTFM_SCROBBLER(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST ((obj), MAFW_LASTFM_TYPE_SCROBBLER, MafwLastfmScrobbler))

#define MAFW_LASTFM_SCROBBLER_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST ((klass), MAFW_LASTFM_TYPE_SCROBBLER, MafwLastfmScrobblerClass))

#define MAFW_LASTFM_IS_SCROBBLER(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE ((obj), MAFW_LASTFM_TYPE_SCROBBLER))

#define MAFW_LASTFM_IS_SCROBBLER_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE ((klass), MAFW_LASTFM_TYPE_SCROBBLER))

#define MAFW_LASTFM_SCROBBLER_GET_CLASS(obj) \
  (G_TYPE_INSTANCE_GET_CLASS ((obj), MAFW_LASTFM_TYPE_SCROBBLER, MafwLastfmScrobblerClass))

typedef struct MafwLastfmScrobblerPrivate MafwLastfmScrobblerPrivate;

typedef struct {
  GObject parent;
  MafwLastfmScrobblerPrivate *priv;
} MafwLastfmScrobbler;

typedef struct {
  GObjectClass parent_class;
} MafwLastfmScrobblerClass;

typedef struct {
  gchar *artist;
  gchar *title;
  gchar *album;
  glong timestamp;
  gchar source;
  gint64 length;
  gint number;
} MafwLastfmTrack;

GType
mafw_lastfm_scrobbler_get_type (void);

MafwLastfmScrobbler *
mafw_lastfm_scrobbler_new (void);

void
mafw_lastfm_scrobbler_set_credentials (MafwLastfmScrobbler *scrobbler,
                                       const gchar *username,
                                       const gchar *passwd);

void
mafw_lastfm_scrobbler_handshake (MafwLastfmScrobbler *scrobbler);

void
mafw_lastfm_scrobbler_set_playing_now (MafwLastfmScrobbler *scrobbler,
                                       MafwLastfmTrack *track);

void
mafw_lastfm_scrobbler_enqueue_scrobble (MafwLastfmScrobbler *scrobbler,
                                        MafwLastfmTrack *track,
                                        gint position);

void
mafw_lastfm_scrobbler_flush_queue (MafwLastfmScrobbler *scrobbler);

void
mafw_lastfm_scrobbler_suspend (MafwLastfmScrobbler *scrobbler);

MafwLastfmTrack *
mafw_lastfm_track_new (void);

void
mafw_lastfm_track_free (MafwLastfmTrack *track);

G_END_DECLS

#endif /* MAFW_LASTFM_SCROBBLER_H */


control-panel/mafw-lastfm-cpa.c
/**
 * mafw-lastfm: a last.fm scrobbler for mafw
 *
 * Copyright (C) 2009-2010  Claudio Saavedra <csaavedra@igalia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <hildon-cp-plugin/hildon-cp-plugin-interface.h>
#include <hildon/hildon.h>
#include <gtk/gtk.h>
#include <glib.h>
#include <locale.h>
#include <glib/gi18n.h>
#include "config.h"

#define MAFW_LASTFM_CREDENTIALS_FILE ".osso/mafw-lastfm"

typedef struct {
	GtkEntry *username;
	GtkEntry *password;
	gchar *settings_file;
} SettingsContext;

static gchar*
load_username (const gchar *file)
{
	GKeyFile *keyfile;
	gchar *username;

	keyfile = g_key_file_new ();
	g_key_file_load_from_file (keyfile, file, G_KEY_FILE_NONE, NULL);
	username = g_key_file_get_string (keyfile,
					  "Credentials", "username", NULL);

	g_key_file_free (keyfile);

	return username;
}

static void
save_credentials (const gchar *file,
		  const gchar *username,
		  const gchar *password)
{
	gchar *md5passwd;
	GKeyFile *keyfile;

	keyfile = g_key_file_new ();
	md5passwd = g_compute_checksum_for_string (G_CHECKSUM_MD5,
						   password, -1);

	g_key_file_set_string (keyfile, "Credentials",
			       "username", username);
	g_key_file_set_string (keyfile, "Credentials",
			       "password", md5passwd);

	g_file_set_contents (file,
			     g_key_file_to_data (keyfile, NULL, NULL),
			     -1, NULL);
	g_free (md5passwd);
	g_key_file_free (keyfile);
}

static void
on_dialog_response (GtkDialog *dialog,
		    gint response_id,
		    gpointer user_data)
{
	SettingsContext *ctx;
	gchar *username;
	const gchar *password;

	ctx = (SettingsContext *) user_data;

	if (response_id == GTK_RESPONSE_OK)
	{
		username = g_strstrip (g_strdup (gtk_entry_get_text (ctx->username)));
		password = gtk_entry_get_text (ctx->password);

		if (username[0] == '\0' || password[0] == '\0')
		{
			hildon_banner_show_information (GTK_WIDGET (dialog), NULL,
							_("Enter both your username "
							  "and password before saving."));
			return;
		} else
		{
			save_credentials (ctx->settings_file, username, password);
		}
		g_free (username);
	}

	gtk_widget_hide (GTK_WIDGET (dialog));
	gtk_widget_destroy (GTK_WIDGET (dialog));

	g_free (ctx->settings_file);
	g_free (ctx);
}

osso_return_t
execute(osso_context_t *osso, gpointer data, gboolean user_activated)
{
	GtkWidget *dialog;
	GtkWidget *username;
	GtkWidget *password;
	GtkWidget *table;
	GtkWidget *label_username;
	GtkWidget *label_password;
	gchar *usr, *settings_file;
	SettingsContext *ctx;

	setlocale(LC_ALL, "");
	textdomain(GETTEXT_PACKAGE);

	settings_file = g_build_filename (g_get_home_dir (),
					  MAFW_LASTFM_CREDENTIALS_FILE, NULL);

	dialog = gtk_dialog_new_with_buttons (_("Last.fm settings"),
					      GTK_WINDOW (data),
					      GTK_DIALOG_MODAL | GTK_DIALOG_NO_SEPARATOR,
					      /* NB! Leave untranslated. */
					      dgettext ("hildon-libs", "wdgt_bd_done"),
					      GTK_RESPONSE_OK,
					      NULL);

	table = gtk_table_new (2, 2, TRUE);
	label_username = gtk_label_new (_("Username:"));
	gtk_misc_set_alignment (GTK_MISC (label_username), 0.0, 0.5);
	username = hildon_entry_new (HILDON_SIZE_AUTO | HILDON_SIZE_FINGER_HEIGHT);

	gtk_table_attach_defaults (GTK_TABLE (table), label_username, 0, 1, 0, 1);

	gtk_table_attach_defaults (GTK_TABLE (table), username, 1, 2, 0, 1);
	label_password = gtk_label_new (_("Password:"));
	gtk_misc_set_alignment (GTK_MISC (label_password), 0.0, 0.5);
	password = hildon_entry_new (HILDON_SIZE_AUTO | HILDON_SIZE_FINGER_HEIGHT);
	hildon_gtk_entry_set_input_mode (GTK_ENTRY (password),
					 HILDON_GTK_INPUT_MODE_FULL |
					 HILDON_GTK_INPUT_MODE_INVISIBLE);
	gtk_table_attach_defaults (GTK_TABLE (table), label_password, 0, 1, 1, 2);
	gtk_table_attach_defaults (GTK_TABLE (table), password, 1, 2, 1, 2);

	if (g_file_test (settings_file, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_REGULAR))
	{
		usr = load_username (settings_file);
		gtk_entry_set_text (GTK_ENTRY (username), usr);
		gtk_editable_select_region (GTK_EDITABLE (username), 0, -1);
		g_free (usr);
	}

	gtk_container_add (GTK_CONTAINER
			   (gtk_dialog_get_content_area (GTK_DIALOG (dialog))),
			   table);

	ctx = g_new0 (SettingsContext, 1);

	ctx->username = GTK_ENTRY (username);
	ctx->password = GTK_ENTRY (password);
	ctx->settings_file = settings_file;

	g_signal_connect (dialog, "response",
			  G_CALLBACK (on_dialog_response),
			  ctx);

	gtk_widget_show_all (dialog);

	return OSSO_OK;
}

osso_return_t
save_state(osso_context_t *osso, gpointer data)
{
        return OSSO_OK;
}
